File: clinica/backend/app/schemas/paciente.py
from pydantic import BaseModel, EmailStr, validator
from typing import Optional
from datetime import date, datetime
from models.paciente import GeneroEnum, EstadoCivilEnum, TipoSangreEnum

class PacienteBase(BaseModel):
    # Datos personales
    nombres: str
    apellidos: str
    fecha_nacimiento: date
    genero: GeneroEnum
    dpi: str
    
    # Contacto
    telefono_principal: str
    telefono_secundario: Optional[str] = None
    email: Optional[EmailStr] = None
    direccion: str
    municipio: Optional[str] = None
    departamento: Optional[str] = None
    
    # Información adicional
    estado_civil: Optional[EstadoCivilEnum] = None
    religion: Optional[str] = None
    ocupacion: Optional[str] = None
    tipo_sangre: Optional[TipoSangreEnum] = None
    
    # Seguro
    tiene_igss: bool = False
    numero_igss: Optional[str] = None
    tiene_seguro_privado: bool = False
    nombre_seguro: Optional[str] = None
    
    # Emergencia
    emergencia_nombre: str
    emergencia_telefono: str
    emergencia_parentesco: Optional[str] = None
    
    # Otros
    observaciones: Optional[str] = None

    @validator('dpi')
    def validar_dpi(cls, v):
        """Validar formato de DPI guatemalteco (13 dígitos)"""
        if not v.isdigit():
            raise ValueError('DPI debe contener solo números')
        if len(v) != 13:
            raise ValueError('DPI debe tener exactamente 13 dígitos')
        return v
    
    @validator('telefono_principal', 'telefono_secundario', 'emergencia_telefono')
    def validar_telefono(cls, v):
        """Validar formato de teléfono guatemalteco"""
        if v is None:
            return v
        # Remover espacios y guiones
        telefono = v.replace(' ', '').replace('-', '')
        if not telefono.isdigit():
            raise ValueError('Teléfono debe contener solo números')
        if len(telefono) != 8:
            raise ValueError('Teléfono debe tener 8 dígitos')
        return telefono

class PacienteCreate(PacienteBase):
    pass

class PacienteUpdate(PacienteBase):
    nombres: Optional[str] = None
    apellidos: Optional[str] = None
    fecha_nacimiento: Optional[date] = None
    genero: Optional[GeneroEnum] = None
    dpi: Optional[str] = None
    telefono_principal: Optional[str] = None
    direccion: Optional[str] = None
    emergencia_nombre: Optional[str] = None
    emergencia_telefono: Optional[str] = None

class PacienteResponse(PacienteBase):
    id: int
    activo: bool
    foto_url: Optional[str]
    edad: int
    created_at: datetime
    updated_at: Optional[datetime]

    class Config:
        from_attributes = True

class PacienteListItem(BaseModel):
    """Para listar pacientes (menos campos)"""
    id: int
    nombres: str
    apellidos: str
    dpi: str
    telefono_principal: str
    edad: int
    genero: GeneroEnum
    foto_url: Optional[str]
    activo: bool

    class Config:
        from_attributes = True

File: clinica/frontend/types/paciente.ts
// types/paciente.ts

// Enums que coinciden con el backend
export enum GeneroEnum {
  MASCULINO = "Masculino",
  FEMENINO = "Femenino",
  OTRO = "Otro"
}

export enum EstadoCivilEnum {
  SOLTERO = "Soltero/a",
  CASADO = "Casado/a",
  DIVORCIADO = "Divorciado/a",
  VIUDO = "Viudo/a",
  UNION_LIBRE = "Unión Libre"
}

export enum TipoSangreEnum {
  A_POSITIVO = "A+",
  A_NEGATIVO = "A-",
  B_POSITIVO = "B+",
  B_NEGATIVO = "B-",
  AB_POSITIVO = "AB+",
  AB_NEGATIVO = "AB-",
  O_POSITIVO = "O+",
  O_NEGATIVO = "O-"
}

export enum CategoriaArchivoEnum {
  FOTO = "foto",
  LABORATORIO = "laboratorio",
  IMAGEN = "imagen",
  RECETA = "receta",
  EKG = "ekg",
  VIDEO = "video",
  OTRO = "otro"
}

// Interface base de Paciente
export interface PacienteBase {
  // Datos personales
  nombres: string;
  apellidos: string;
  fecha_nacimiento: string; // ISO date string
  genero: GeneroEnum;
  dpi: string;
  
  // Contacto
  telefono_principal: string;
  telefono_secundario?: string;
  email?: string;
  direccion: string;
  municipio?: string;
  departamento?: string;
  
  // Información adicional
  estado_civil?: EstadoCivilEnum;
  religion?: string;
  ocupacion?: string;
  tipo_sangre?: TipoSangreEnum;
  
  // Seguro
  tiene_igss: boolean;
  numero_igss?: string;
  tiene_seguro_privado: boolean;
  nombre_seguro?: string;
  
  // Emergencia
  emergencia_nombre: string;
  emergencia_telefono: string;
  emergencia_parentesco?: string;
  
  // Otros
  observaciones?: string;
}

// Para crear paciente
export interface PacienteCreate extends PacienteBase {}

// Para actualizar paciente (todos opcionales)
export interface PacienteUpdate extends Partial<PacienteBase> {
  activo?: boolean;
}

// Respuesta completa del servidor
export interface Paciente extends PacienteBase {
  id: number;
  activo: boolean;
  foto_url?: string;
  created_at: string;
}

// Para listar pacientes (vista reducida)
export interface PacienteListItem {
  id: number;
  nombres: string;
  apellidos: string;
  dpi: string;
  telefono_principal: string;
  genero: GeneroEnum;
  foto_url?: string;
  activo: boolean;
}

// Archivos del paciente
export interface ArchivoPaciente {
  id: number;
  paciente_id: number;
  categoria: CategoriaArchivoEnum;
  nombre_archivo: string;
  ruta_archivo: string;
  tipo_mime: string;
  tamano_bytes: number;
  descripcion?: string;
  created_at: string;
}

// Estadísticas
export interface EstadisticasPacientes {
  total_pacientes: number;
  pacientes_activos: number;
  pacientes_inactivos: number;
  nuevos_este_mes: number;
  por_genero: Record<string, number>;
  por_tipo_sangre: Record<string, number>;
}

// Helper para calcular edad
export function calcularEdad(fechaNacimiento: string): number {
  const hoy = new Date();
  const nacimiento = new Date(fechaNacimiento);
  let edad = hoy.getFullYear() - nacimiento.getFullYear();
  const mes = hoy.getMonth() - nacimiento.getMonth();
  
  if (mes < 0 || (mes === 0 && hoy.getDate() < nacimiento.getDate())) {
    edad--;
  }
  
  return edad;
}

// Helper para formatear nombre completo
export function nombreCompleto(paciente: { nombres: string; apellidos: string }): string {
  return `${paciente.nombres} ${paciente.apellidos}`;
}

// Helper para formatear DPI
export function formatearDPI(dpi: string): string {
  // Formato: 1234 56789 0123
  return dpi.replace(/(\d{4})(\d{5})(\d{4})/, '$1 $2 $3');
}

// Helper para formatear teléfono
export function formatearTelefono(telefono: string): string {
  // Formato: 1234-5678
  return telefono.replace(/(\d{4})(\d{4})/, '$1-$2');
}